searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp

```diff
@@ -220,23 +223,45 @@ DenseTensorAttribute::onLoad(vespalib::Executor *executor)
     uint32_t numDocs(tensorReader.getDocIdLimit());
     _refVector.reset();
     _refVector.unsafe_reserve(numDocs);
+    auto complete_executor = (_index && !use_index_file && executor)
+            ? std::make_unique<vespalib::ThreadStackExecutor>(1, 0x10000)
+            : std::unique_ptr<vespalib::ThreadStackExecutor>();
+    std::atomic<uint64_t> pending(0);
     for (uint32_t lid = 0; lid < numDocs; ++lid) {
         if (tensorReader.is_present()) {
             auto raw = _denseTensorStore.allocRawBuffer();
             tensorReader.readTensor(raw.data, _denseTensorStore.getBufSize());
             _refVector.push_back(raw.ref);
             if (_index && !use_index_file) {
-                // This ensures that get_vector() (via getTensor()) is able to find the newly added tensor.
-                setCommittedDocIdLimit(lid + 1);
-                _index->add_document(lid);
-                if ((lid % 256) == 0) {
-                    commit();
+                if (executor != nullptr) {
+                    while (pending > 1000) { std::this_thread::yield(); }
```

Comment:
Just yield causes wasted cpu usage on systems with many CPU cores. Consider using a condition variable.

Reply:
Refactored and fixed. PTAL